### PR TITLE
Rewrite calls to `std::ptr::addr_of` and `std::ptr::addr_of_mut`

### PR DESCRIPTION
**Pull request by aneksteind**
- [x] *mut
- [x] *const

Pointers with `CELL` permissions will be supported in another PR

### CODE REVIEW

File: c2rust-analyze/src/rewrite/expr/hir_op.rs [hunk 1]
```diff
@@ -159,6 +159,23 @@ impl<'a, 'tcx> HirRewriteVisitor<'a, 'tcx> {
                     .unwrap_or_else(|err| panic_sole_location_error(err, "Assign statement"));
                 opt_assign_loc
             }
+            hir::ExprKind::AddrOf(..) => {
+                if ex.span == ex.span.source_callsite() {
+                    // FIXME: this is a hacky proxy for checking if &raw comes from
```
Comment (aneksteind) [inline c2rust-analyze/src/rewrite/expr/hir_op.rs:164]: this is addressed in https://github.com/immunant/c2rust/pull/893/files#diff-f27494e760407f9ad0a4266c21cf7c2e553c775034319d545a0fba0e58989965L163

**Review by fw-immunant — APPROVED** [review]
LGTM, makes sense.

I don't like how the rewrites become `&(x)` when we do try to keep track of precedence, but I believe that's caused by `Rewrite::Sub` always being parenthesized (as is also the case for `Rewrite::Identity`) rather than by anything in this PR.

**Comment by aneksteind** [comment]
@spernsteiner would you mind taking a look at this?

File: c2rust-analyze/src/rewrite/expr/hir_op.rs [hunk 2]
```diff
@@ -221,6 +238,12 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirRewriteVisitor<'a, 'tcx> {
     fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
         let mut hir_rw = Rewrite::Identity;
 
+        // This span will be used to apply the actual rewrite.
+        // To prevent attempts to rewrite macros in their definition
+        // location (e.g. `std::ptr::addr_of`) we instead specify that
+        // the rewrite should occur at the callsite
+        let callsite_span = ex.span.source_callsite();
```
Comment (spernsteiner) [inline c2rust-analyze/src/rewrite/expr/hir_op.rs:245]: Always looking through one level of macro expansion is probably not the right answer in all situations, but should work for the code we get out of `c2rust-transpile` at least.

**Review by spernsteiner — APPROVED** [review]

**Pull request merged**